Build failure: grafana on aarch64-darwin

### Steps To Reproduce

Steps to reproduce the behavior:

1. `nix run nixpkgs#grafana` on aarch64-darwin

### Build log

```
$ nix run -L nixpkgs#grafana
grafana> unpacking sources
grafana> unpacking source archive /nix/store/0ncdnawnzfzllb291ci8khkkhrwh4vdv-source
grafana> source root is source
grafana> patching sources
grafana> updateAutotoolsGnuConfigScriptsPhase
grafana> configuring
grafana> building
grafana> wire: github.com/grafana/grafana/pkg/server: wrote /private/tmp/nix-build-grafana-10.2.2.drv-0/source/pkg/server/wire_gen.go
grafana> Building subPackage .
grafana> Building subPackage ./.github/workflows/scripts/kinds
grafana> Building subPackage ./kinds
grafana> package github.com/grafana/grafana/kinds: build constraints exclude all Go files in /private/tmp/nix-build-grafana-10.2.2.drv-0/source/kinds
grafana> Building subPackage ./pkg
grafana> package github.com/grafana/grafana/pkg: build constraints exclude all Go files in /private/tmp/nix-build-grafana-10.2.2.drv-0/source/pkg
grafana> Building subPackage ./pkg/api
grafana> pkg/tsdb/mssql/mssql.go:20:2: cannot find module providing package github.com/microsoft/go-mssqldb: import lookup disabled by -mod=vendor
grafana> pkg/tsdb/mssql/mssql.go:21:2: cannot find module providing package github.com/microsoft/go-mssqldb/azuread: import lookup disabled by -mod=vendor
error: builder for '/nix/store/acvann56zagva3jni5mg8yfwxrfn1vrf-grafana-10.2.2.drv' failed with exit code 1;
       last 17 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/0ncdnawnzfzllb291ci8khkkhrwh4vdv-source
       > source root is source
       > patching sources
       > updateAutotoolsGnuConfigScriptsPhase
       > configuring
       > building
       > wire: github.com/grafana/grafana/pkg/server: wrote /private/tmp/nix-build-grafana-10.2.2.drv-0/source/pkg/server/wire_gen.go
       > Building subPackage .
       > Building subPackage ./.github/workflows/scripts/kinds
       > Building subPackage ./kinds
       > package github.com/grafana/grafana/kinds: build constraints exclude all Go files in /private/tmp/nix-build-grafana-10.2.2.drv-0/source/kinds
       > Building subPackage ./pkg
       > package github.com/grafana/grafana/pkg: build constraints exclude all Go files in /private/tmp/nix-build-grafana-10.2.2.drv-0/source/pkg
       > Building subPackage ./pkg/api
       > pkg/tsdb/mssql/mssql.go:20:2: cannot find module providing package github.com/microsoft/go-mssqldb: import lookup disabled by -mod=vendor
       > pkg/tsdb/mssql/mssql.go:21:2: cannot find module providing package github.com/microsoft/go-mssqldb/azuread: import lookup disabled by -mod=vendor
       For full logs, run 'nix log /nix/store/acvann56zagva3jni5mg8yfwxrfn1vrf-grafana-10.2.2.drv'.
```

### Additional context


### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.1.0, macOS 14.1.2`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.1`
 - channels(bouke): `""`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

@Ma27 If I just add the darwin cache for yarn, it fails at:

```
 > Building subPackage ./pkg/cmd/grafana
       > pkg/tsdb/mssql/mssql.go:22:2: cannot find module providing package github.com/microsoft/go-mssqldb: import lookup disabled by -mod=vendor
       > pkg/tsdb/mssql/mssql.go:23:2: cannot find module providing package github.com/microsoft/go-mssqldb/azuread: import lookup disabled by -mod=vendor
```

Which is the same issue as reported by OP. To solve this, I had used `proxyVendor = true;`. When I use this the `go-modules` no longer have the `github.com/microsoft\~nix\~case\~hack\~1` anymore, but there is `github.com/microsoft` and `github.com/!microsoft`.  I am guessing `proxyVendor` somehow works around the case-insensitivity problem.

--- a/pkgs/servers/monitoring/grafana/default.nix
+++ b/pkgs/servers/monitoring/grafana/default.nix
@@ -2,7 +2,7 @@
 , tzdata, wire
 , yarn, nodejs, python3, cacert
 , jq, moreutils
-, nix-update-script, nixosTests
+, nix-update-script, nixosTests, xcbuild
 }:
 
 let
@@ -32,13 +32,21 @@ buildGoModule rec {
     hash = "sha256-Rp2jGspbmqJFzSbiVy2/5oqQJnAdGG/T+VNBHVsHSwg=";
   };
 
+  # borrowed from: https://github.com/NixOS/nixpkgs/blob/d70d9425f49f9aba3c49e2c389fe6d42bac8c5b0/pkgs/development/tools/analysis/snyk/default.nix#L20-L22
+  env = lib.optionalAttrs (stdenv.isDarwin && stdenv.isx86_64) {
+    # Fix error: no member named 'aligned_alloc' in the global namespace.
+    # Occurs while building @esfx/equatable@npm:1.0.2 on x86_64-darwin
+    NIX_CFLAGS_COMPILE = "-D_LIBCPP_HAS_NO_LIBRARY_ALIGNED_ALLOCATION=1";
+  };
+
   offlineCache = stdenv.mkDerivation {
     name = "${pname}-${version}-yarn-offline-cache";
-    inherit src;
+    inherit src env;
     nativeBuildInputs = [
       yarn nodejs cacert
       jq moreutils python3
-    ];
+    # @esfx/equatable@npm:1.0.2 fails to build on darwin as it requires `xcbuild`
+    ] ++ lib.optionals stdenv.isDarwin [ xcbuild.xcbuild ];
     postPatch = ''
       ${patchAwayGrafanaE2E}
     '';
@@ -47,7 +55,7 @@ buildGoModule rec {
       export HOME="$(mktemp -d)"
       yarn config set enableTelemetry 0
       yarn config set cacheFolder $out
-      yarn config set --json supportedArchitectures.os '[ "linux" ]'
+      yarn config set --json supportedArchitectures.os '[ "linux", "darwin" ]'
       yarn config set --json supportedArchitectures.cpu '["arm", "arm64", "ia32", "x64"]'
       yarn
       runHook postBuild
@@ -56,14 +64,21 @@ buildGoModule rec {
     dontInstall = true;
     dontFixup = true;
     outputHashMode = "recursive";
-    outputHash = "sha256-QdyXSPshzugkDTJoUrJlHNuhPAyR9gae5Cbk8Q8FSl4=";
+    outputHash = rec {
+      x86_64-linux = "sha256-3CZgs732c6Z64t2sfWjPAmMFKVTzoolv2TwrbjeRCBA=";
+      aarch64-linux = x86_64-linux;
+      aarch64-darwin = "sha256-NKEajOe9uDZw0MF5leiKBIRH1CHUELRho7gyCa96BO8=";
+      x86_64-darwin = aarch64-darwin;
+    }.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
   };
 
   disallowedRequisites = [ offlineCache ];
 
-  vendorHash = "sha256-cNkMVLXp3hPMcW0ilOM0VlrrDX/IsZaze+/6qlTfmRs=";
+  vendorHash = "sha256-puPgbgfRqbPvMVks+gyOPOTTfdClWqbOf89X0ihMLPY=";
 
-  nativeBuildInputs = [ wire yarn jq moreutils removeReferencesTo python3 ];
+  proxyVendor = true;
+
+  nativeBuildInputs = [ wire yarn jq moreutils removeReferencesTo python3 ] ++ lib.optionals stdenv.isDarwin [ xcbuild.xcbuild ];
 
   postPatch = ''
     ${patchAwayGrafanaE2E}

--- a/pkgs/servers/monitoring/grafana/default.nix
+++ b/pkgs/servers/monitoring/grafana/default.nix
@@ -152,7 +152,7 @@ buildGoModule rec {
     license = licenses.agpl3;
     homepage = "https://grafana.com";
     maintainers = with maintainers; [ offline fpletz willibutz globin ma27 Frostman ];
-    platforms = platforms.linux ++ platforms.darwin;
+    platforms = [ "x86_64-linux" "x86_64-darwin" "aarch64-linux" "aarch64-darwin" ];
     mainProgram = "grafana-server";
   };
 }

